Cassandra 18129 livemigration eviction

aweisberg opened the pull request.

aweisberg committed:
Accord support for live migration

Random checkpoint

foo

random checkpoint

random checkpoint

Random checkpoint

Cleanup checkpoint

Cleanup before review

Handle errors in BeginRecovery, mark Defer as transient so it's not lost

Accord isn't really properly simulated, but at least this allows epoch updates to work.

Remove getRejectBefore

Make setting result methods not visible on Barrer

Suppress unused warnings on loggers

Update TODOs and make it an error for ExistingTransactionCheck to find multiple candidate transactions for a key

Replace todo with comment explaining rationale for surfacing invalidation as timeout

Factor out ProposeSyncPoint from CoordinateSyncPoint and fix several test bugs in CoordinateTest

Add awaitility as a test dependency, switch AsyncChain from Executor to ExecutorService, some other random review feedback

Remove redundant AsyncResult.addCallback methods which are default methods in the AsyncResult interface

Add shutdown check for executors in AsyncCallbacks

Add Command.addAndInvokeListener to make it easier to check and wait on commands using the listeners

Allow multiple keys/ranges for global barriers

Fix deadlock in burn test causes by waiting on an AsyncChain

aweisberg committed:
Fix imports

bdeggleston commented on accord-core/src/main/java/accord/local/SafeCommand.java:
```diff
@@ -71,6 +70,12 @@ public Command addListener(Command.DurableAndIdempotentListener listener)
         return update(Command.addListener(current(), listener));
     }
 
+    public void addAndInvokeListener(SafeCommandStore safeStore, TransientListener listener)
```
could you move this into the InMemorySafeCommand so it's not part of the public api?

aweisberg replied:
I think either we should remove this and make callers manually invoke the listener or starting using it in `Barrier` in which case it should be part of the API. I don't really have a preference, but for now I am going to switch `Barrier` to use it.

bdeggleston commented on accord-core/src/main/java/accord/coordinate/Barrier.java:
```diff
@@ -0,0 +1,309 @@
+            BarrierTxn found = safeStore.mapReduceWithTerminate(
+                    seekables,
+                    safeStore.ranges().allSince(minEpoch),
+                    // TODO review this with Benedict, but I don't believe an exclusive sync point works here
```
is this TODO still relevant?

aweisberg replied:
@belliottsmith For a barrier where we want to know that everything from before that transaction has applied already would exclusive sync points work?

I think they would because they wouldn't enter the `Applied` state until their dependencies were applied and they do take the necessary dependencies?

belliottsmith replied:
Define before? An `ExclusiveSyncPoint` defines an order on `txnId` not `executeAt` - their `executeAt` is sort of imprecise, somewhere after its own `txnId` and its dependency with the highest `executeAt` - logically different on each replica and key.

But once they apply, anything with a lower `txnId` has applied, yes. It sounds to me like you might want a NoOp transaction though, which finds an `executeAt` that is after any conflicting transaction that had started before it starts.

aweisberg replied:
`local` barrier (this path) only need to establish that no transactions that execute in epochs < `minEpoch` are unapplied. It's OK if transactions that execute in the current epoch or later are not applied. `local` barriers will try to use an existing transaction to establish that the constraint is satisfied and if they can't they create a `SyncPoint` transaction.

`global` barriers need to establish a happens before relationship between any observed side effects of transactions from before the creation of the barrier, and the application of those transactions. The two use cases are ensuring no accord transactions from an older epoch are unapplied (migration) and ensuring that any transaction side effects witnessed by repair are from now committed transactions. Global barriers always create a `SyncPoint` transaction and then wait on it.

If a `SyncPoint` and `ExclusiveSyncPoint` both have the property that once they are applied than any transaction with a lower `txnId` has applied then they should both be viable transaction types to wait on instead of creating a new `SyncPoint` transaction.

When searching for a transaction the parameters provided are `TestTimestamp.EXECUTES_AFTER` and `TxnId.minForEpoch(minEpoch)` so I thought that meant I was picking a transaction with an established `executeAt` that once applied also means the application of any transaction with a `txnId` or `executeAt` <  the selected transaction and that it would work with read and read/write transactions.

belliottsmith replied:
> once applied also means the application of any transaction with a txnId or executeAt < the selected transaction

By definition, txnId < executeAt.

> If a SyncPoint and ExclusiveSyncPoint both have the property that once they are applied than any transaction with a lower txnId has applied

Only an ExclusiveSyncPoint has this property, as a lower `txnId` may not have yet agreed to execute when the `SyncPoint` was agreed, and the standard `SyncPoint` does not invalidate the earlier `txnId`. However, such a `txnId` would execute logically after the `SyncPoint`. We don't make any specific guarantees about `executeAt` because we don't synchronise across the whole range simultaneously, it's a partial order. But I think this satisfies your required properties.

bdeggleston commented on accord-core/src/main/java/accord/coordinate/Barrier.java:
```diff
@@ -0,0 +1,309 @@
+                    TxnId.minForEpoch(minEpoch),
+                    TestDep.ANY_DEPS,
+                    null,
+                    // TODO is PreApplied right? It's the first OutcomeKnown state, but why not Committed?
```
is this TODO still relevant?

aweisberg replied:
I still have this question. I think it's correct I am just not 100% on why. Is `Committed` not good enough because `Committed` doesn't mean the outcome is known because it could be `Committed` at some nodes and `Invalidated` at others?

belliottsmith replied:
No, if something is `Committed` it cannot be `Invalidated`. The state machine branches at `PreCommitted` - from here it must go to `Applied`. We otherwise skip straight from `AcceptedInvalidated` (at a quorum) to `Invalidated` which is a terminal state.

aweisberg replied:
Ah so I could use `Committed` instead? I don't want to use anything that could still be `Invalidated` because then it might not wait on its dependencies.

I must have misunderstood what `OutcomeKnown` encompasses.

belliottsmith replied:
Outcome refers to the `Writes` and `Result` registers, which are computed during the Execute phase.

bdeggleston commented on accord-core/src/main/java/accord/coordinate/Barrier.java:
```diff
@@ -0,0 +1,309 @@
+                RoutingKey homeKey = syncPoint.homeKey;
+                node.commandStores().ifLocal(contextFor(txnId), homeKey, epoch, epoch, safeStore -> {
+                    Command.TransientListener listener = new BarrierCommandListener();
+                    // TODO handle null return?
```
is this TODO still relevant?

aweisberg replied:
This one isn't

bdeggleston commented on accord-core/src/main/java/accord/coordinate/Barrier.java:
```diff
@@ -0,0 +1,309 @@
+            if (found != null && !found.status.equals(Status.Applied))
+            {
+                safeStore.commandStore().execute(contextFor(found.txnId), safeStoreWithTxn -> {
+                    // TODO handle null return from get due to truncation
```
is this TODO still relevant?

aweisberg replied:
I don't think it is, it won't return null right? It will return something non-null for the transaction id saying it's not known or truncated.

bdeggleston reviewed (changes requested):
there are a lot of TODOs that needs to either be addressed or removed. I commented on a few, but I counted 9-10 total

aweisberg committed:
Fix as much rebase/upstream issues as possible without rebasing

aweisberg committed:
Use addAndInvokeListener in Barrier, handle truncation and not just invalidation, replace addListener with addAndInvokeListener in Barrier in one spot that risked missing invoking the listener.

aweisberg committed:
Remove unused delay and cork functionality

aweisberg committed:
Remove some TODOs by reworking/correcting how coordinators/recovery coordinators issue BlockOnDeps

BlockOnDeps can end up getting an insufficient response (not 100% clear on the full range of cases where this can happen). I think it can happen if an earlier part of the transaction is missed by a node like Preaccept so it never got the transaction description?

So modify ApplyAndWaitUntilApplied to always set up the listener even if the Apply fails and respond with not committed.

Then the coordinator (recovery or otherwise) now responds with a maximal Apply (not commit which was a mistake) to enable the application
to proceed.

aweisberg committed:
Yes that slice looks correct to me

aweisberg committed:
Small whitespace change

aweisberg committed:
Remove debug output

aweisberg committed:
Allow Committed to transactions to be listened to for local barriers beacuse they won't be invalidated and should eventually complete.
